File: 0207-course-schedule/0207-course-schedule.py
"""
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # can finish current course's prerequisites?
        self.finished = set()
        
        for i in range(numCourses):
            if not self.canFinishRecursive(i, prerequisites):
                return False
        return True   
    
    def canFinishRecursive(self, course: int, prerequisites: List[List[int]]) -> bool:
        if course in self.finished:
            return True
        
        for p in prerequisites:
            if p[0] == course:
                if not self.canFinishRecursive(p[1], prerequisites):
                    return False
        
        self.finished.add(course)       
        return True

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        self.finished = set()
        for i in range(numCourses):
            if self.contains_cycle(i, set(), prerequisites):
                return False
        return True
    
    def contains_cycle(self, course: int, visited: Set[int], prerequisites: List[List[int]]) -> bool:
        if course in self.finished:
            return False
        if course in visited:
            return True
        visited.add(course)
        for p in prerequisites:    
            if course == p[0]:
                if self.contains_cycle(p[1], visited, prerequisites):
                    return True
                    
        self.finished.add(course)            
        return False            
        
"""

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # represent prerequisites as a graph dict k, v 
        # where k is a vertex and v is a list of edges
        graph = {}
        # vertex to incoming edges
        requires = {}
        for p in prerequisites:
            graph[p[1]] = graph.get(p[1], []) + [p[0]]
            requires[p[0]] = requires.get(p[0], 0) + 1
        
        queue = []
        for vertex in graph.keys():
            print(vertex)
            if requires.get(vertex, 0) == 0:
                queue.append(vertex)
        
        while queue:
            cur = queue.pop()
            for e in graph.get(cur, []):
                requires[e] -= 1
                if requires[e] == 0:
                    queue.append(e)
        
        for edges in requires.values():
            if edges != 0:
                return False
        
        return True
        
        

File: 0207-course-schedule/0207-course-schedule.go
/*
func canFinish(numCourses int, prerequisites [][]int) bool {
    if len(prerequisites) == 0 {
        return true
    }
    
    finishable := make(map[int]bool, numCourses)
    for i := 0; i < numCourses; i++ {
        visited := make(map[int]bool, numCourses)
        if !canFinishRecursive(i, visited, finishable, prerequisites) {
            return false
        }
    }
    
    return true
}


func canFinishRecursive(course int, visited, finishable map[int]bool, prerequisites [][]int) bool {
    if finishable[course] {
        return true
    }
    
    if visited[course] {
        return false
    }
    
    visited[course] = true
    for _, p := range prerequisites {
        if p[0] == course {
            if !canFinishRecursive(p[1], visited, finishable, prerequisites) {
                return false
            }
        }
    }
    
    finishable[course] = true
    return true
}
*/

func canFinish(numCourses int, prerequisites [][]int) bool {
    // prereqFor maps each course to a slice of
    // courses for which it is a prerequisite
    prereqFor := make(map[int][]int, numCourses)
    
    // prereqCount maps a course to its 
    // number of prerequisites
    prereqCount := make(map[int]int, numCourses)
    
    // initialize the declared graph and map
    for _, prereq := range prerequisites {
        course, req := prereq[0], prereq[1]
        prereqFor[req] = append(prereqFor[req], course)
        prereqCount[course]++
    }
    
    // initialize a queue with valid starting courses,
    // i.e. courses with no prerequisites
    queue := make([]int, 0, numCourses)
    for course, _ := range prereqFor {
        if prereqCount[course] == 0 {
            queue = append(queue, course)
        }
    }
    
    // bfs through courses, appending courses only 
    // after their prerequisites have been visited
    for len(queue) != 0 {
        req := queue[0]
        queue = queue[1:]
        for _, course := range prereqFor[req] {
            prereqCount[course]--
            if prereqCount[course] == 0 {
                queue = append(queue, course)
            }
        }
    }
    
    // check that each course had its 
    // prerequisites met during bfs traversal
    for _, count := range prereqCount {
        if count != 0 {
            return false
        }
    }
    return true
}